moodlecli/utils.py

```diff
@@ -238,3 +238,101 @@ def maybe_user_uuids(moodle_client, user_ids=[]):
             return []
         else:
             raise e
+
+
+def update_grades_data(moodle_client, course_id, old_grades):
+    """This utility function builds a single object which includes data from
+    multiple Moodle endpoints. The final object includes the following:
+    {
+        "usergrades": [ // Data as returned by get_grades_by_course ],
+        "quizzes": [ // Data as returned by get_quizzes_by_courses ],
+        "attempts": {
+            "{user_id}": {
+                "{quiz_id}": {
+                    "summaries": [
+                        // Data as returned by get_user_quiz_attempts
+                    ]
+                    "details": {
+                        "{attempt_id}": {
+                            // Data as returned by get_quiz_attempt_details
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    The function expects the same structure when parsing to determine if
+    data from old_grades can be used as a cache to avoid unnecessary calls
+    to Moodle.
+    """
+
+    def _attempts_data_is_stale(attempts, user_id, quiz_id, new_attempt_ts):
+        old_user_attempts = attempts.get(user_id, {})
+        old_quiz_attempts = old_user_attempts.get(quiz_id, {})
+
+        # Use the gradednotificationsenttime to decide whether data is stale
+        latest_found_ts = 0
+        for summary in old_quiz_attempts.get("summaries", []):
+            tmp_attempt_ts = summary["gradednotificationsenttime"]
+            if (tmp_attempt_ts is not None) and \
+               (latest_found_ts < tmp_attempt_ts):
+                latest_found_ts = tmp_attempt_ts
+
+        return new_attempt_ts > latest_found_ts
+
+    # Always pull latest grades and quiz data. The grades data is the basis
+    # of the data object returned by this function.
+    new_grades = moodle_client.get_grades_by_course(course_id)
+    new_quizzes = moodle_client.get_quizzes_by_courses([course_id])
+
+    new_grades["quizzes"] = new_quizzes["quizzes"]
+    new_grades["attempts"] = {}
+
+    new_usergrades = new_grades["usergrades"]
+    old_attempts = old_grades.get("attempts", {})
+
+    for new_usergrade in new_usergrades:
```

Comment:
@P-Gill97 - I'm definitely open to thoughts on how we might improve this to be more readable.